File: native/mlx-sidecar/Sources/lrgenius-mlx/Protocol.swift
// The wire protocol between the Rust server (`lrg-mlx`) and this sidecar.
//
// One JSON object per line, requests on stdin and responses on stdout, each
// response echoing the request's `id`. Requests are served strictly in order:
// MLX evaluation is not reentrant here and a photo batch saturates the GPU
// anyway, so there is nothing to gain from interleaving.
//
// stdout carries *only* protocol lines. Anything the process wants to say to a
// human goes to stderr, which the Rust side relays into the server log. Getting
// this wrong corrupts the stream, so nothing in this target may `print`.
//
// Images do not travel in the JSON. A decoded photo is several megabytes of raw
// RGB and base64 would inflate every batch line past comfort, so the Rust side
// writes the pixels to a temp file and sends the path; see `ImagePayload`.

import Foundation

// MARK: - Requests

/// The operations the sidecar understands. Decoded via `op`.
enum RequestOp: String, Codable {
    case ping
    case load
    case generate
    case status
    case unload
    case shutdown
}

struct SidecarRequest: Decodable {
    let id: Int
    let op: RequestOp

    /// `load` only: the directory holding an MLX model (config.json, weights,
    /// tokenizer). Always a local path — the sidecar never downloads anything,
    /// because the Rust side owns download progress reporting for every other
    /// model in the app and a second, invisible download path would be worse
    /// than no download path.
    let modelDir: String?

    /// `generate` only.
    let requests: [GenerationSpec]?

    enum CodingKeys: String, CodingKey {
        case id
        case op
        case modelDir = "model_dir"
        case requests
    }
}

/// Raw pixels for one photo, staged in a file the Rust side owns.
struct ImagePayload: Decodable {
    let path: String
    let width: Int
    let height: Int
    /// Bytes per pixel in `path`: 3 for RGB, 4 for RGBA. Only 3 is sent today;
    /// carried explicitly so the reader never has to infer it from the length.
    let channels: Int
}

/// One photo's worth of work. Mirrors `lrg_providers::local::LocalRequest`.
struct GenerationSpec: Decodable {
    let systemPrompt: String
    /// Run-constant across an indexing batch. Kept as its own field rather than
    /// pre-joined so the ordering that makes prefix reuse possible is decided
    /// here and not accidentally reshuffled by a caller.
    let stablePrompt: String
    /// Varies per photo.
    let perPhotoPrompt: String
    let image: ImagePayload?
    /// A JSON Schema, already serialized. Decoding is constrained to it when
    /// present and free when absent.
    let schema: String?
    let maxTokens: Int
    let temperature: Float

    enum CodingKeys: String, CodingKey {
        case systemPrompt = "system_prompt"
        case stablePrompt = "stable_prompt"
        case perPhotoPrompt = "per_photo_prompt"
        case image
        case schema
        case maxTokens = "max_tokens"
        case temperature
    }
}

// MARK: - Responses

/// What the engine settled on after loading. Mirrors `lrg_llama::EngineInfo`
/// closely enough that `/llm/status` can report either backend uniformly.
struct EngineInfoPayload: Encodable {
    let modelPath: String
    let modelName: String
    let supportsVision: Bool

    enum CodingKeys: String, CodingKey {
        case modelPath = "model_path"
        case modelName = "model_name"
        case supportsVision = "supports_vision"
    }
}

/// One photo's result. `ok == false` carries `error` and nothing else, so one
/// bad photo never sinks the rest of a batch.
struct GenerationResultPayload: Encodable {
    let ok: Bool
    let text: String?
    let promptTokens: Int?
    let completionTokens: Int?
    /// Always false on this backend, for two reasons that both have to be
    /// fixed before it could ever be true.
    ///
    /// `GuidedGenerationLoop.run` allocates a fresh KV cache per call
    /// (`model.newCache`) and neither accepts nor returns one, so there is no
    /// seam to hand it a pre-filled cache. Still true at mlx-swift-lm HEAD as
    /// of 2026-08-13; the prompt-cache work in #475 landed in MLXLMCommon, not
    /// in the guided loop.
    ///
    /// And even with that seam there would be almost nothing to reuse: Gemma's
    /// message generator orders the user turn as image-then-text, so the photo
    /// precedes the run-constant prompt and only the system turn is common
    /// across photos. See the ordering note in `Engine.generateOne`.
    ///
    /// Reported honestly rather than optimistically: the llama.cpp backend
    /// really does pin its prefix, and conflating the two would make the
    /// `--debug` logs lie about where the time goes.
    let prefixReused: Bool
    let error: String?

    enum CodingKeys: String, CodingKey {
        case ok
        case text
        case promptTokens = "prompt_tokens"
        case completionTokens = "completion_tokens"
        case prefixReused = "prefix_reused"
        case error
    }

    static func success(text: String, promptTokens: Int, completionTokens: Int) -> Self {
        .init(
            ok: true, text: text, promptTokens: promptTokens,
            completionTokens: completionTokens, prefixReused: false, error: nil)
    }

    static func failure(_ message: String) -> Self {
        .init(
            ok: false, text: nil, promptTokens: nil, completionTokens: nil,
            prefixReused: false, error: message)
    }
}

struct SidecarResponse: Encodable {
    let id: Int
    let ok: Bool
    let error: String?
    let info: EngineInfoPayload?
    let results: [GenerationResultPayload]?
    /// `status` only: whether a model is currently resident.
    let loaded: Bool?

    static func ok(id: Int) -> Self {
        .init(id: id, ok: true, error: nil, info: nil, results: nil, loaded: nil)
    }

    static func failure(id: Int, _ message: String) -> Self {
        .init(id: id, ok: false, error: message, info: nil, results: nil, loaded: nil)
    }

    static func loaded(id: Int, info: EngineInfoPayload) -> Self {
        .init(id: id, ok: true, error: nil, info: info, results: nil, loaded: true)
    }

    static func status(id: Int, info: EngineInfoPayload?) -> Self {
        .init(id: id, ok: true, error: nil, info: info, results: nil, loaded: info != nil)
    }

    static func generated(id: Int, results: [GenerationResultPayload]) -> Self {
        .init(id: id, ok: true, error: nil, info: nil, results: results, loaded: nil)
    }
}

File: server-rs/crates/lrg-mlx/src/protocol.rs
//! The wire types shared with `native/mlx-sidecar/Sources/lrgenius-mlx/Protocol.swift`.
//!
//! Both sides are hand-written rather than generated, so the field names here
//! are the contract: change one and you must change the other. The Swift side
//! spells every multi-word key in snake_case via explicit `CodingKeys` so these
//! structs need no rename attributes beyond the ones below.

use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize)]
pub struct Request<'a> {
    pub id: u64,
    pub op: &'static str,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model_dir: Option<&'a str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub requests: Option<Vec<GenerationSpec>>,
}

/// Where the raw pixels for one photo were staged.
///
/// Pixels travel via a file rather than inside the JSON: a decoded photo is
/// several megabytes and base64 would add a third again to every batch line.
#[derive(Debug, Serialize)]
pub struct ImagePayload {
    pub path: String,
    pub width: u32,
    pub height: u32,
    /// 3 for RGB, 4 for RGBA. Always 3 today; sent explicitly so the reader
    /// never infers it from the file length.
    pub channels: u32,
}

#[derive(Debug, Serialize)]
pub struct GenerationSpec {
    pub system_prompt: String,
    pub stable_prompt: String,
    pub per_photo_prompt: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub image: Option<ImagePayload>,
    /// The schema is sent pre-serialized because XGrammar compiles from schema
    /// *text*; round-tripping it through a Swift dictionary and back would risk
    /// reordering keys and defeating the sidecar's grammar cache.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub schema: Option<String>,
    pub max_tokens: u32,
    pub temperature: f32,
}

#[derive(Debug, Deserialize)]
pub struct EngineInfoPayload {
    pub model_path: String,
    pub model_name: String,
    pub supports_vision: bool,
}

#[derive(Debug, Deserialize)]
pub struct GenerationResultPayload {
    pub ok: bool,
    pub text: Option<String>,
    pub prompt_tokens: Option<u32>,
    pub completion_tokens: Option<u32>,
    #[serde(default)]
    pub prefix_reused: bool,
    pub error: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct Response {
    pub id: u64,
    pub ok: bool,
    pub error: Option<String>,
    pub info: Option<EngineInfoPayload>,
    pub results: Option<Vec<GenerationResultPayload>>,
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The Swift side keys off exactly these names; a rename here is a silent
    /// protocol break, so pin them.
    #[test]
    fn generation_spec_uses_the_agreed_key_names() {
        let spec = GenerationSpec {
            system_prompt: "sys".into(),
            stable_prompt: "stable".into(),
            per_photo_prompt: "photo".into(),
            image: Some(ImagePayload {
                path: "/tmp/x.rgb".into(),
                width: 2,
                height: 3,
                channels: 3,
            }),
            schema: Some("{}".into()),
            max_tokens: 64,
            temperature: 0.2,
        };
        let json = serde_json::to_value(&spec).unwrap();
        for key in [
            "system_prompt",
            "stable_prompt",
            "per_photo_prompt",
            "image",
            "schema",
            "max_tokens",
            "temperature",
        ] {
            assert!(json.get(key).is_some(), "missing key {key}");
        }
        assert_eq!(json["image"]["channels"], 3);
    }

    /// An absent image must be omitted, not sent as `null`: the Swift decoder
    /// treats the key's presence as "there is a photo here".
    #[test]
    fn an_absent_image_is_omitted_entirely() {
        let spec = GenerationSpec {
            system_prompt: String::new(),
            stable_prompt: String::new(),
            per_photo_prompt: String::new(),
            image: None,
            schema: None,
            max_tokens: 8,
            temperature: 0.0,
        };
        let json = serde_json::to_value(&spec).unwrap();
        assert!(json.get("image").is_none());
        assert!(json.get("schema").is_none());
    }

    #[test]
    fn a_response_without_optional_fields_still_decodes() {
        let response: Response = serde_json::from_str(r#"{"id":7,"ok":true}"#).unwrap();
        assert_eq!(response.id, 7);
        assert!(response.ok);
        assert!(response.results.is_none());
    }
}
